File: DesktopHub/src/DesktopHub.Infrastructure/Firebase/FirebaseAuth.cs
using System.Net.Http;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using DesktopHub.Infrastructure.Logging;

namespace DesktopHub.Infrastructure.Firebase;

/// <summary>
/// Owns the client-side Firebase Auth lifecycle.
///
/// Flow:
///   1. POST to the `issueToken` Cloud Function with {licenseKey, username, deviceId}
///      — returns a short-lived custom token whose claims carry the caller's tier.
///   2. Exchange that custom token for an ID token via the Identity Toolkit
///      `signInWithCustomToken` REST endpoint.
///   3. Cache the ID token and its expiry. All RTDB REST calls then append
///      `?auth=&lt;idToken&gt;` which Firebase evaluates against the database rules.
///
/// The client NEVER holds any service-account credentials. The only "secret"
/// it needs is the Web API key, which is public by design (identifies the
/// Firebase project and is shipped in every web/mobile Firebase app).
/// </summary>
public sealed class FirebaseAuth : IDisposable
{
    // The Web API key is public information — visible in the Firebase console
    // under Project Settings → General → Web API Key. It identifies the project
    // for public APIs like Identity Toolkit. It does NOT grant any privileged
    // access; all writes are still gated by custom-token claims + RTDB rules.
    private const string WebApiKey = "AIzaSyBq0-qnsspK-FOm-SvCBgBbBHL_Tx1k8WI";

    private const string IssueTokenUrl =
        "https://us-central1-licenses-ff136.cloudfunctions.net/issueToken";

    private const string SignInWithCustomTokenUrl =
        "https://identitytoolkit.googleapis.com/v1/accounts:signInWithCustomToken?key=" + WebApiKey;

    private const string RefreshTokenUrl =
        "https://securetoken.googleapis.com/v1/token?key=" + WebApiKey;

    // Refresh a bit before the token technically expires to avoid racing.
    private static readonly TimeSpan RefreshGrace = TimeSpan.FromMinutes(5);

    private readonly HttpClient _http;
    private readonly SemaphoreSlim _gate = new(1, 1);

    private string? _idToken;
    private string? _refreshToken;
    private DateTime _expiresAtUtc = DateTime.MinValue;
    private string? _tier;
    private string? _username;
    private string? _licenseKey;
    private string? _deviceId;

    public FirebaseAuth()
    {
        _http = new HttpClient { Timeout = TimeSpan.FromSeconds(15) };
        _http.DefaultRequestHeaders.Add("User-Agent", "DesktopHub-Client/1.0");
    }

    public string? Tier => _tier;
    public string? Username => _username;
    public bool IsReady => !string.IsNullOrEmpty(_idToken) && DateTime.UtcNow < _expiresAtUtc;

    /// <summary>Initial handshake. Idempotent — re-calling before expiry is a no-op.</summary>
    public async Task<bool> SignInAsync(string licenseKey, string username, string deviceId)
    {
        _licenseKey = licenseKey;
        _username = username.ToLowerInvariant();
        _deviceId = deviceId;

        await _gate.WaitAsync().ConfigureAwait(false);
        try
        {
            if (IsReady) return true;
            return await DoSignInLocked().ConfigureAwait(false);
        }
        finally { _gate.Release(); }
    }

    /// <summary>
    /// Returns a valid Firebase ID token, refreshing transparently if needed.
    /// Returns null if we don't have a refresh token yet (i.e. SignInAsync
    /// was never called or failed).
    /// </summary>
    public async Task<string?> GetIdTokenAsync()
    {
        await _gate.WaitAsync().ConfigureAwait(false);
        try
        {
            if (!string.IsNullOrEmpty(_idToken) && DateTime.UtcNow + RefreshGrace < _expiresAtUtc)
                return _idToken;

            // Prefer cheap refresh if we have one; otherwise do full sign-in again.
            if (!string.IsNullOrEmpty(_refreshToken))
            {
                if (await TryRefreshLocked().ConfigureAwait(false))
                    return _idToken;
            }

            if (!string.IsNullOrEmpty(_licenseKey) && !string.IsNullOrEmpty(_username) &&
                !string.IsNullOrEmpty(_deviceId))
            {
                await DoSignInLocked().ConfigureAwait(false);
                return _idToken;
            }

            return null;
        }
        finally { _gate.Release(); }
    }

    private async Task<bool> DoSignInLocked()
    {
        try
        {
            // 1. Call issueToken — Firebase v2 callable format: POST { "data": {...} }
            var issueBody = JsonSerializer.Serialize(new
            {
                data = new
                {
                    licenseKey = _licenseKey,
                    username = _username,
                    deviceId = _deviceId
                }
            });

            using var issueResp = await _http.PostAsync(
                IssueTokenUrl,
                new StringContent(issueBody, Encoding.UTF8, "application/json")).ConfigureAwait(false);

            if (!issueResp.IsSuccessStatusCode)
            {
                var body = await issueResp.Content.ReadAsStringAsync().ConfigureAwait(false);
                InfraLogger.Log($"FirebaseAuth: issueToken failed {(int)issueResp.StatusCode}: {body}");
                return false;
            }

            var issueJson = await issueResp.Content.ReadAsStringAsync().ConfigureAwait(false);
            using var issueDoc = JsonDocument.Parse(issueJson);
            if (!issueDoc.RootElement.TryGetProperty("result", out var issueResult))
            {
                InfraLogger.Log("FirebaseAuth: issueToken response missing 'result'");
                return false;
            }

            var customToken = issueResult.GetProperty("token").GetString();
            _tier = issueResult.TryGetProperty("tier", out var tierEl) ? tierEl.GetString() : "user";

            if (string.IsNullOrEmpty(customToken))
            {
                InfraLogger.Log("FirebaseAuth: empty custom token from issueToken");
                return false;
            }

            // 2. Exchange the custom token for an ID token via Identity Toolkit
            var exchangeBody = JsonSerializer.Serialize(new
            {
                token = customToken,
                returnSecureToken = true
            });

            using var exchResp = await _http.PostAsync(
                SignInWithCustomTokenUrl,
                new StringContent(exchangeBody, Encoding.UTF8, "application/json")).ConfigureAwait(false);

            if (!exchResp.IsSuccessStatusCode)
            {
                var body = await exchResp.Content.ReadAsStringAsync().ConfigureAwait(false);
                InfraLogger.Log($"FirebaseAuth: signInWithCustomToken failed {(int)exchResp.StatusCode}: {body}");
                return false;
            }

            var exchJson = await exchResp.Content.ReadAsStringAsync().ConfigureAwait(false);
            using var exchDoc = JsonDocument.Parse(exchJson);
            var root = exchDoc.RootElement;

            _idToken = root.GetProperty("idToken").GetString();
            _refreshToken = root.TryGetProperty("refreshToken", out var rt) ? rt.GetString() : null;
            var expiresIn = int.TryParse(
                root.TryGetProperty("expiresIn", out var ex) ? ex.GetString() : "3600",
                out var secs) ? secs : 3600;
            _expiresAtUtc = DateTime.UtcNow.AddSeconds(expiresIn);

            InfraLogger.Log($"FirebaseAuth: signed in as '{_username}' tier={_tier} (expires in {expiresIn}s)");
            return !string.IsNullOrEmpty(_idToken);
        }
        catch (Exception ex)
        {
            InfraLogger.Log($"FirebaseAuth: SignIn threw: {ex.Message}");
            return false;
        }
    }

    private async Task<bool> TryRefreshLocked()
    {
        try
        {
            var body = $"grant_type=refresh_token&refresh_token={Uri.EscapeDataString(_refreshToken!)}";
            using var resp = await _http.PostAsync(
                RefreshTokenUrl,
                new StringContent(body, Encoding.UTF8, "application/x-www-form-urlencoded")).ConfigureAwait(false);

            if (!resp.IsSuccessStatusCode)
            {
                var errBody = await resp.Content.ReadAsStringAsync().ConfigureAwait(false);
                InfraLogger.Log($"FirebaseAuth: refresh failed {(int)resp.StatusCode}: {errBody}");
                return false;
            }

            var json = await resp.Content.ReadAsStringAsync().ConfigureAwait(false);
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            _idToken = root.GetProperty("id_token").GetString();
            _refreshToken = root.TryGetProperty("refresh_token", out var rt) ? rt.GetString() : _refreshToken;
            var expiresIn = int.TryParse(
                root.TryGetProperty("expires_in", out var ex) ? ex.GetString() : "3600",
                out var secs) ? secs : 3600;
            _expiresAtUtc = DateTime.UtcNow.AddSeconds(expiresIn);

            return !string.IsNullOrEmpty(_idToken);
        }
        catch (Exception ex)
        {
            InfraLogger.Log($"FirebaseAuth: refresh threw: {ex.Message}");
            return false;
        }
    }

    /// <summary>
    /// Forwards a Cloud Functions callable invocation with the current ID token
    /// as the Bearer credential. Used by the developer panel for admin ops
    /// (pushForceUpdate, clearForceUpdate).
    /// </summary>
    public async Task<JsonElement?> CallFunctionAsync(string functionName, object payload)
    {
        var idToken = await GetIdTokenAsync().ConfigureAwait(false);
        if (string.IsNullOrEmpty(idToken))
        {
            InfraLogger.Log($"FirebaseAuth: CallFunction({functionName}) aborted — no ID token");
            return null;
        }

        var url = $"https://us-central1-licenses-ff136.cloudfunctions.net/{functionName}";
        var bodyJson = JsonSerializer.Serialize(new { data = payload });

        using var req = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = new StringContent(bodyJson, Encoding.UTF8, "application/json")
        };
        req.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", idToken);

        using var resp = await _http.SendAsync(req).ConfigureAwait(false);
        var respBody = await resp.Content.ReadAsStringAsync().ConfigureAwait(false);

        if (!resp.IsSuccessStatusCode)
        {
            InfraLogger.Log($"FirebaseAuth: {functionName} failed {(int)resp.StatusCode}: {respBody}");
            return null;
        }

        try
        {
            using var doc = JsonDocument.Parse(respBody);
            if (doc.RootElement.TryGetProperty("result", out var result))
                return result.Clone();
        }
        catch { /* malformed — treat as failure */ }

        return null;
    }

    public void Dispose()
    {
        _http.Dispose();
        _gate.Dispose();
    }
}

File: Renamer/src/firebase_auth.py
"""Client-side Firebase Auth lifecycle for Spec Header Updater.

Mirrors the pattern used by DesktopHub's FirebaseAuth.cs:

  1. POST to the `issueToken` Cloud Function with
     {appId, licenseKey, username, deviceId}. The function validates the
     license, resolves the caller's tier (user / dev / admin) from
     Firebase role nodes, and returns a short-lived (1-hour) Firebase
     custom token whose claims carry the tier + app identity.
  2. Exchange the custom token for a Firebase ID token via the Identity
     Toolkit `signInWithCustomToken` REST endpoint.
  3. Cache the ID token + refresh token. Refresh transparently a few
     minutes before expiry.

All subsequent Realtime Database REST calls attach the ID token as
`?auth=<idToken>`. The server enforces tier- and app-scoped rules
against the token's claims -- the client therefore holds *zero*
service-account credentials and cannot do anything a non-admin user
isn't already authorized to do.

The Web API key hardcoded below is public information (it identifies
the Firebase project for public APIs like Identity Toolkit) and is
NOT a credential. See README / docs for background.
"""
from __future__ import annotations

import json
import threading
import time
import urllib.parse
import urllib.request
from datetime import datetime, timedelta, timezone
from typing import Any, Dict, Optional


# Web API key for the licenses-ff136 Firebase project. Public by design --
# shipped in every Firebase web/mobile app. Grants no privileged access on
# its own; writes are still gated by custom-token claims + database rules.
WEB_API_KEY = "AIzaSyBTmqZ6HMdKEoR8bMtSka-gUm_XiqwLYHM"

ISSUE_TOKEN_URL = (
    "https://us-central1-licenses-ff136.cloudfunctions.net/issueToken"
)
SIGN_IN_WITH_CUSTOM_TOKEN_URL = (
    "https://identitytoolkit.googleapis.com/v1/accounts:signInWithCustomToken"
    f"?key={WEB_API_KEY}"
)
REFRESH_TOKEN_URL = (
    f"https://securetoken.googleapis.com/v1/token?key={WEB_API_KEY}"
)

# Refresh this many seconds BEFORE the token's stated expiry to avoid racing
# with a server-side clock skew or network latency.
REFRESH_GRACE_SECONDS = 5 * 60

_HTTP_TIMEOUT_SECONDS = 15


class FirebaseAuthError(RuntimeError):
    """Raised when the client cannot obtain or refresh a Firebase ID token."""


class FirebaseAuth:
    """Owns the sign-in -> refresh cycle for a single app session.

    Thread-safe via a single lock; call GetIdToken from any thread.
    """

    def __init__(self, app_id: str):
        self._app_id = app_id
        self._lock = threading.Lock()

        self._id_token: Optional[str] = None
        self._refresh_token: Optional[str] = None
        self._expires_at_utc: datetime = datetime.fromtimestamp(0, tz=timezone.utc)
        self._tier: Optional[str] = None
        self._user_id: Optional[str] = None   # HMAC hash from issueToken claim
        self._tenant_id: Optional[str] = None

        # Remembered args so we can re-sign-in if the refresh path fails.
        self._license_key: Optional[str] = None
        self._username: Optional[str] = None
        self._device_id: Optional[str] = None

    # ----- public API ------------------------------------------------------

    @property
    def tier(self) -> Optional[str]:
        return self._tier

    @property
    def user_id(self) -> Optional[str]:
        """Hashed user id issued by the Cloud Function. Use for every
        user-keyed RTDB path. Raw Windows username never leaves memory."""
        return self._user_id

    @property
    def tenant_id(self) -> Optional[str]:
        return self._tenant_id

    @property
    def is_ready(self) -> bool:
        return bool(self._id_token) and datetime.now(timezone.utc) < self._expires_at_utc

    def sign_in(self, license_key: str, username: str, device_id: str,
                tenant_id: str) -> bool:
        """Initial handshake. Idempotent -- re-calling before expiry is a no-op."""
        with self._lock:
            self._license_key = license_key
            self._username = username.lower()
            self._device_id = device_id
            self._tenant_id = tenant_id

            if self.is_ready:
                return True
            return self._do_sign_in_locked()

    def get_id_token(self) -> Optional[str]:
        """Return a valid ID token, refreshing transparently if needed.

        Returns None when we don't have prior context (sign_in was never
        called, or failed, and we have no refresh token cached).
        """
        with self._lock:
            now = datetime.now(timezone.utc)
            if self._id_token and (now + timedelta(seconds=REFRESH_GRACE_SECONDS)) < self._expires_at_utc:
                return self._id_token

            # Prefer a cheap refresh when we have a refresh_token; fall back
            # to a fresh sign-in otherwise.
            if self._refresh_token and self._try_refresh_locked():
                return self._id_token

            if self._license_key and self._username and self._device_id:
                if self._do_sign_in_locked():
                    return self._id_token

            return None

    def call_function(self, function_name: str, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Invoke a Firebase callable-formatted function with the current ID token.

        Callers that need to reach admin-tier Cloud Functions
        (e.g. pushForceUpdate) use this; the request is authenticated by the
        caller's tier claim, the function then checks req.auth.token.tier
        server-side before executing any privileged operation.
        """
        id_token = self.get_id_token()
        if not id_token:
            return None

        url = f"https://us-central1-licenses-ff136.cloudfunctions.net/{function_name}"
        body = {"data": payload}
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {id_token}",
        }
        try:
            resp_json = _http_post_json(url, body, headers=headers)
        except Exception as exc:
            print(f"FirebaseAuth.call_function({function_name}) failed: {exc}")
            return None

        return resp_json.get("result") if isinstance(resp_json, dict) else None

    # ----- locked internals ------------------------------------------------

    def _do_sign_in_locked(self) -> bool:
        """Full round-trip: issueToken -> signInWithCustomToken.

        Caller already holds self._lock.
        """
        try:
            issue_body = {
                "data": {
                    "appId": self._app_id,
                    "licenseKey": self._license_key,
                    "username": self._username,
                    "deviceId": self._device_id,
                    "tenantId": self._tenant_id,
                }
            }
            issue_resp = _http_post_json(
                ISSUE_TOKEN_URL, issue_body,
                headers={"Content-Type": "application/json"},
            )
            result = issue_resp.get("result") if isinstance(issue_resp, dict) else None
            if not result:
                print(f"FirebaseAuth.sign_in: issueToken response missing 'result': {issue_resp!r}")
                return False

            custom_token = result.get("token")
            self._tier = result.get("tier", "user")
            self._user_id = result.get("userId")
            if result.get("tenantId"):
                self._tenant_id = result.get("tenantId")
            if not custom_token:
                print("FirebaseAuth.sign_in: empty custom token from issueToken")
                return False

            # Exchange the custom token for a Firebase ID token.
            exch_resp = _http_post_json(
                SIGN_IN_WITH_CUSTOM_TOKEN_URL,
                {"token": custom_token, "returnSecureToken": True},
                headers={"Content-Type": "application/json"},
            )
            self._id_token = exch_resp.get("idToken")
            self._refresh_token = exch_resp.get("refreshToken")
            expires_in = int(exch_resp.get("expiresIn", 3600))
            self._expires_at_utc = datetime.now(timezone.utc) + timedelta(seconds=expires_in)

            # Never log raw username -- user_id is the non-reversible identifier.
            print(
                f"FirebaseAuth: signed in tenant={self._tenant_id} "
                f"user_id={self._user_id} tier={self._tier} "
                f"(expires in {expires_in}s)"
            )
            return bool(self._id_token)
        except Exception as exc:
            print(f"FirebaseAuth.sign_in: {type(exc).__name__}: {exc}")
            return False

    def _try_refresh_locked(self) -> bool:
        """Swap the refresh_token for a new id_token (cheaper than a full sign-in)."""
        try:
            form = urllib.parse.urlencode({
                "grant_type": "refresh_token",
                "refresh_token": self._refresh_token,
            }).encode()
            req = urllib.request.Request(
                REFRESH_TOKEN_URL, data=form, method="POST",
                headers={"Content-Type": "application/x-www-form-urlencoded"},
            )
            with urllib.request.urlopen(req, timeout=_HTTP_TIMEOUT_SECONDS) as r:
                body = json.loads(r.read().decode("utf-8") or "{}")

            self._id_token = body.get("id_token")
            self._refresh_token = body.get("refresh_token", self._refresh_token)
            expires_in = int(body.get("expires_in", 3600))
            self._expires_at_utc = datetime.now(timezone.utc) + timedelta(seconds=expires_in)
            return bool(self._id_token)
        except Exception as exc:
            print(f"FirebaseAuth.refresh: {type(exc).__name__}: {exc}")
            return False


# ----- helpers -----------------------------------------------------------


def _http_post_json(url: str, body: Dict[str, Any], *,
                     headers: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
    data = json.dumps(body).encode("utf-8")
    req = urllib.request.Request(
        url, data=data, method="POST",
        headers=headers or {"Content-Type": "application/json"},
    )
    with urllib.request.urlopen(req, timeout=_HTTP_TIMEOUT_SECONDS) as r:
        raw = r.read().decode("utf-8")
        return json.loads(raw) if raw else {}
